### PR TITLE
Add a distributed Dataframe API to Ray

### PR DESCRIPTION
**Pull request by devin-petersohn**
## What do these changes do?
Adds a Pandas DataFrame wrapper for use in Ray:

* Distributes the DataFrame.
* Operates on the DataFrame partitions in remote tasks.
* Currently provides some simple functionality.

## TODO
- [x] Add/Format unit tests in test/dataframe.py
- [x] Clean up a couple of quick hacks
- [ ] ~~Add skeleton for full API with `NotImplementedErrors`~~
- [ ] ~~Docs in doc/source~~
- [ ] ~~Benchmark~~

Push the last 3 to next PR.

### CODE REVIEW

**Comment by devin-petersohn** [comment]
I'll ping when the tests are in. Jenkins doesn't need to test until then.

**Comment by pcmoritz** [comment]
ok to test

**Comment by robertnishihara** [comment]
add to whitelist

**Comment by robertnishihara** [comment]
Not a big deal, but the "add to whitelist" command doesn't seem to work for me. I had to go and add @devin-petersohn manually to the jenkins config.

cc @shaneknapp

File: python/ray/dataframe.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
+import pandas as pd
+import numpy as np
+import ray
+
+ray.register_custom_serializer(pd.DataFrame, use_pickle=True)
```
Comment (robertnishihara) [inline python/ray/dataframe.py]: this is to work around #1305?

File: python/ray/dataframe.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
+import pandas as pd
+import numpy as np
+import ray
```
Comment (robertnishihara) [inline python/ray/dataframe.py]: I think it makes sense to move the file to `python/ray/dataframe/dataframe.py`. And to add a file `python/ray/dataframe/__init__.py` that has some lines like

```
from .dataframe import DataFrame
```

**Review by robertnishihara — COMMENTED** [review]
Want to add some simple tests for this? E.g., make a file `test/dataframe_test.py` and call it from `.travis.yml`.

Note that most of our tests use `unittest` but I think we'd like to switch to `pytest`, so if you make a new test file, let's use `pytest`.

File: python/ray/dataframe.py (hunk 1 above)
Comment (devin-petersohn) [reply python/ray/dataframe.py]: Yes, it turns out the `pd.core.indexes.base.Index` was also giving problems.

File: .travis.yml [hunk 3]
```diff
@@ -122,6 +122,7 @@ script:
   - python test/trial_runner_test.py
   - python test/trial_scheduler_test.py
   - python test/cython_test.py
+  - python -m pytest test/dataframe.py
```
Comment (devin-petersohn) [inline .travis.yml]: Testing through `pytest`.

**Comment by robertnishihara** [comment]
Can you take a look at the Travis test failures at https://travis-ci.org/ray-project/ray/builds/317618386?utm_source=github_status&utm_medium=notification?

Looks like we need to install pandas on Travis. The files to modify are `doc/requirements.txt` and `.travis/install-dependencies.sh`.

**Comment by pcmoritz** [comment]
You will also need to solve the rebase conflict, otherwise Travis doesn't run.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import pandas as pd
+import numpy as np
+import ray
+
+ray.register_custom_serializer(pd.DataFrame, use_pickle=True)
+ray.register_custom_serializer(pd.core.indexes.base.Index, use_pickle=True)
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: now that this is in `__init__.py`, we can probably remove it from here

File: python/ray/dataframe/__init__.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from .dataframe import DataFrame
```
Comment (robertnishihara) [inline python/ray/dataframe/__init__.py]:
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

File: test/dataframe.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
+import pytest
```
Comment (robertnishihara) [inline test/dataframe.py]: also add the "from future" stuff here

File: python/ray/dataframe/__init__.py (hunk 5 above)
Comment (devin-petersohn) [reply python/ray/dataframe/__init__.py]: Sorry about that! Please see above.

File: test/dataframe.py (hunk 6 above)
Comment (devin-petersohn) [reply test/dataframe.py]: Added above.

**Comment by pcmoritz** [comment]
There are a few listing errors: https://travis-ci.org/ray-project/ray/jobs/318924011

Do you want to fix the linting and then we get the PR merged and work on the TODOs you mention above in the next PR?

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**